cli.py

```diff
@@ -86,3 +87,80 @@ def case_search(query: str, num_cases: int):
+@click.argument("bookmarks", nargs=-1, type=int)
+@click.option('-n', '--num-cases', default=5, show_default=True, help='Maximum number of recommmendation results')
+@click.option('-c', '--court', multiple=True, default=[], help='Filter to a specific court id (can be provided multiple times)')
+def case_recommend(bookmarks: Tuple[int], num_cases: int, court: Tuple[str]):
+    from algorithms import CaseRecommendation
+    from extraction.citation_extractor import CitationExtractor
+    from graph import CitationNetwork
+    from db.peewee.models import Opinion, Cluster
+    citation_network = CitationNetwork.get_citation_network(enable_caching=True)
+    recommendation = CaseRecommendation(citation_network)
+    recommendations = recommendation.recommendations(frozenset(bookmarks), num_cases, courts=frozenset(court))
+    with get_session() as s:
+        ro = sorted(
+            Opinion.select().join(Cluster).where(Opinion.resource_id << list(recommendations.keys())),
+            key=lambda op: recommendations[op.resource_id],
+            reverse=True
+        )
+        for op in ro:
+            click.echo(pretty_print_opinion(op))
+
+def case_recall(cases: (Tuple[int], int), court: Tuple[str], numtrials: int, samecourt: bool):
+    from algorithms import CaseRecommendation
+    from extraction.citation_extractor import CitationExtractor
+    from graph import CitationNetwork
+    from db.peewee.models import Opinion, Cluster
+    import random
+    from numpy.random import choice
+    citation_network = CitationNetwork.get_citation_network(enable_caching=True)
+    recommendation = CaseRecommendation(citation_network)
+    if isinstance(cases, int):
+        # cases will be selected in proportion to how often they're mentioned --- this seems good
+        cases = choice(citation_network.network_edge_list.edge_list, size=cases)
+    overall_top1 = 0
+    overall_top5 = 0
+    overall_top20 = 0
+    for c in cases:
+        md = citation_network.network_edge_list.node_metadata[c]
+        neighbors = list(citation_network.network_edge_list.edge_list[md.start:md.end])
+        top1 = 0
+        top5 = 0
+        top20 = 0
+        for i in range(numtrials):
+            removed = neighbors.pop(random.randrange(len(neighbors)))
+            same_court = (citation_network.network_edge_list.node_metadata[removed].court,) if samecourt else ()
+            recommendations = list(recommendation.recommendations(frozenset(neighbors), 20, courts=frozenset(court + same_court), ignore_opinion_ids=frozenset([c])).keys())
+            if removed in recommendations:
+                top20 += 1
+            if removed in recommendations[:5]:
+                top5 += 1
+            if removed == recommendations[0]:
+                top1 += 1
+            neighbors.append(removed)
+        overall_top1 += top1
+        overall_top5 += top5
+        overall_top20 += top20
```

Comment:
I'm actually not sure if this is a correct definition of recall (see https://chaitanyabelhekar.medium.com/recommender-system-metrics-clearly-explained-1f2ba6690216) but I need to do some more research into this to understand what other papers in our area do.

Reply:
Satisfied that for our purposes this is correct.